File: src/ObjectManager/Object.Tes/FilePacks/Records/345-LTEX.Land Texture.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class LTEXRecord : Record, IHaveEDID
    {
        public struct HNAMField
        {
            public byte MaterialType;
            public byte Friction;
            public byte Restitution;
           
            public HNAMField(UnityBinaryReader r, int dataSize)
            {
                MaterialType = r.ReadByte();
                Friction = r.ReadByte();
                Restitution = r.ReadByte();
            }
        }

        public override string ToString() => $"LTEX: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public FILEField ICON; // Texture
        // TES3
        public INTVField INTV;
        // TES4
        public HNAMField HNAM; // Havok data
        public BYTEField SNAM; // Texture specular exponent
        public List<FMIDField<GRASRecord>> GNAMs = new List<FMIDField<GRASRecord>>(); // Potential grass

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "INTV": INTV = new INTVField(r, dataSize); return true;
                case "ICON":
                case "DATA": ICON = new FILEField(r, dataSize); return true;
                // TES4
                case "HNAM": HNAM = new HNAMField(r, dataSize); return true;
                case "SNAM": SNAM = new BYTEField(r, dataSize); return true;
                case "GNAM": GNAMs.Add(new FMIDField<GRASRecord>(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-LTEX.Land Texture.swift
//
//  LTEXRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class LTEXRecord: Record, IHaveEDID {
    public typealias HNAMField = (
        materialType: UInt8,
        friction: UInt8,
        restitution: UInt8
    )
    
    public override var description: String { return "LTEX: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var ICON: FILEField! // Texture
    // TES3
    public var INTV: INTVField!
    // TES4
    public var HNAM: HNAMField! // Havok data
    public var SNAM: BYTEField! // Texture specular exponent
    public var GNAMs = [FMIDField<GRASRecord>]() // Potential grass

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "INTV": INTV = r.readINTV(dataSize)
        case "ICON",
             "DATA": ICON = r.readSTRV(dataSize)
        // TES4
        case "HNAM": HNAM = r.readT(dataSize)
        case "SNAM": SNAM = r.readT(dataSize)
        case "GNAM": GNAMs.append(FMIDField<GRASRecord>(r, dataSize))
        default: return false
        }
        return true
    }
}
